modules/core/src/main/scala/sangria/execution/FutureResolver.scala

```diff
@@ -1534,6 +1539,140 @@ private[execution] class FutureResolver[Ctx](
     Result(errorReg, if (canceled) None else Some(marshaller.arrayNode(listBuilder.result())))
   }
 
+  private def trackDeprecation(ctx: Context[Ctx, _]): Unit = {
+    val visitedDirectives = mutable.Set[String]()
+
+    def getArgValue(name: String, args: Args): Option[Input] =
+      if (args.argDefinedInQuery(name)) {
+        if (args.optionalArgs.contains(name)) {
+          args.argOpt(name)
+        } else {
+          Some(args.arg(name))
+        }
+      } else {
+        None
+      }
+
+    def deprecatedArgsUsed(argDefs: List[Argument[_]], argValues: Args): List[Argument[_]] =
+      argDefs.filter { argDef =>
+        val argValue = getArgValue(argDef.name, argValues)
+        argDef.deprecationReason.isDefined && argValue.isDefined && iu.isDefined(argValue.get)
+      }
+
+    def trackDeprecatedDirectiveArgs(astDirective: ast.Directive): Unit = {
+      // prevent infinite loop from directiveA -> arg -> directiveA -> arg ...
+      if (visitedDirectives.contains(astDirective.name)) {
+        return
+      }
+      visitedDirectives.add(astDirective.name)
+
+      ctx.schema.directives.find(_.name == astDirective.name) match {
+        case Some(directive) =>
+          val directiveArgs = valueCollector
+            .getArgumentValues(
+              Some(astDirective),
+              directive.arguments,
+              astDirective.arguments,
+              variables)
+
+          directiveArgs match {
+            case Success(directiveArgs) =>
+              deprecatedArgsUsed(directive.arguments, directiveArgs).foreach { arg =>
+                deprecationTracker.deprecatedDirectiveArgUsed(directive, arg, ctx)
+              }
+            case _ => // if we fail to get args, the query should fail elsewhere
+          }
+
+          // nested argument directives
+          directive.arguments.foreach { nestedArg =>
+            nestedArg.astDirectives.foreach(trackDeprecatedDirectiveArgs)
+          }
+        case _ => // do nothing
+      }
+    }
+
+    def trackDeprecatedInputObjectFields(inputType: InputType[_], ioArg: Input): Unit =
+      inputType match {
+        case ioDef: InputObjectType[_] =>
+          ioDef.fields.foreach { field =>
+            // field deprecation
+            val fieldVal: Option[Input] = (ioArg match {
+              case lm: ListMap[String, Input] => lm.get(field.name)
+              case in: Input =>
```

Comment:
I'm a little confused about how `Args` is set at this point

at first I was assuming it was all `Input` and didn't have the `ListMap` check here (and didn't have the `Seq` check below for lists) but that was causing errors

it doesn't appear to actually be `Input` although everything compiles fine, I had to add a `try { } catch { }` for `None` values being here in some tests

removing the input unmarshaller and `Input` bit entirely seems to actually still work just fine, so think I can just remove that? (going to double check and make a new commit for this)